cohttp-curl/src/cohttp_curl.ml

```diff
@@ -0,0 +1,138 @@
+module Sink = struct
+  type _ t = String : string t | Discard : unit t
+
+  let string = String
+  let discard = Discard
+end
+
+module Source = struct
+  type t = Empty | String of string
+
+  let empty = Empty
+  let string s = String s
+
+  let to_curl_callback t =
+    match t with
+    | Empty -> fun _ -> ""
+    | String s ->
+        let len = String.length s in
+        let pos = ref 0 in
+        fun max_asked ->
+          if !pos >= len then ""
+          else
+            let chunk_len = min (len - !pos) max_asked in
+            let res = String.sub s !pos chunk_len in
+            pos := !pos + chunk_len;
+            res
+end
+
+module Request = struct
+  type 'a t = {
+    curl : Curl.t;
+    body : 'a Sink.t;
+    mutable body_buffer : Buffer.t option;
+  }
+
+  let curl t = t.curl
+
+  let body (type a) (t : a t) : a =
+    match t.body with
+    | Discard ->
+        assert (t.body_buffer = None);
+        ()
+    | String ->
+        let res =
+          Buffer.contents
+            (match t.body_buffer with None -> assert false | Some s -> s)
+        in
+        t.body_buffer <- None;
+        res
+
+  let create (type a) ?timeout_ms ?headers method_ ~uri ~(input : Source.t)
+      ~(output : a Sink.t) ~on_response : a t =
+    let response_header_acc = ref [] in
+    let response_body = ref None in
+    let h = Curl.init () in
+    Curl.setopt h (CURLOPT_URL uri);
+    Curl.setopt h (CURLOPT_CUSTOMREQUEST (Http.Method.to_string method_));
+    let () =
+      match headers with
+      | None -> ()
+      | Some headers ->
+          let buf = Buffer.create 64 in
+          let headers =
+            Http.Header.fold
```

Comment:
Why not using https://github.com/mirage/ocaml-cohttp/blob/master/http/src/http.ml#L174?
If it is faster to use buggers, we could rewrite that, to_frames and to_string to use the buffer

Reply:
I think in a separate pull request it could be interesting to explore using buffers to serialize the response and headers in one go (it’ll help avoid creating some temporary strings) and maybe even explore keeping a long lived buffer per request lifecycle so we don’t need to create a new one when sending a response every time.

Reply:
> Why not using https://github.com/mirage/ocaml-cohttp/blob/master/http/src/http.ml#L174?
If it is faster to use buffers, we could rewrite that, to_frames and to_string to use the buffer

It is because those functions add the `\r\n` which we don't need. I could share the code, but it would require a `Private` module in `Header`.

> I think in a separate pull request it could be interesting to explore using buffers to serialize the response and headers in one go (it’ll help avoid creating some temporary strings) and maybe even explore keeping a long lived buffer per request lifecycle so we don’t need to create a new one when sending a response every time.

Do you mean serialize the request? We don't really serialize responses in the client. Do you mean reuse the buffer between serializing the body and headers?

Reply:
> It is because those functions add the \r\n which we don't need. I could share the code, but it would require a Private module in Header.

I just tried, and there's no good way to do this unfortunately.

Reply:
> It is because those functions add the `\r\n` which we don't need.

The to_frames function does not add them, it should be equivalent to what you have here. The only difference is that it uses sprintf i stead of a buffer